GitUI/UserControls/RevisionGrid/Graph/ThreadSafeGrowingList.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+
+namespace GitUI.UserControls.RevisionGrid.Graph
+{
+    // This list can only grow. Because it only grows, reading is thread safe.
+    // Writing to this list is not thread safe. However, for the usecase
+    // where only one thread is adding items to the list, and multiple threads
+    // are reading, this can be used.
+    public class ThreadSafeGrowingList<T> : IReadOnlyList<T>
+    {
+        private T[] _internalArray;
```

Comment:
Maybe i'm not seeing it, but isn't this more or less an implementation of List<T>? Could we not just use that type? Or at least wrap it?

Reply:
It is the same as the normal list class. We could wrap it, but then we need an extra class layer. This seemed overkill. I also thought about having the list as an private property and only expose the IReadOnlyList<RevisionGraphRevision>. This has the same benefits. So yes, we could do this.

Reply:
If it means we cut down on the code for ThreadSafeGrowingList, then yeah, I'd say go for using List directly or indirectly.

Reply:
Ok, its getting late here, so I cannot do it today. If you want to take a shot at it, please be my guest. You can base it on this branch. I think you can even push directly to this branch on my fork (but maybe its something only collaborators can....).

Reply:
If you don't mind, I'll wait for you to do it. Also quite busy and tired :)

Reply:
🛌 🛌

Reply:
Ok, I tried replaced the low-level array with a list. The 'new class' then simply wraps List<T> and only allows to add items to the internal list. This doesn't work unfortunately. The problem is that if you add an item while another thread is enumerating, you get this error:
System.InvalidOperationException: 'Collection was modified; enumeration operation may not execute.'

The low-level array and the purpose build enumerator class do not have this problem.

What to do next?

Reply:
So the problem is that while enumerating on a List, if you add to it, this can invalidate existing enumerators if List reallocs. In your solution, this doesn't happen because enumerators will keep pointing to the "old" block of memory (and keep it alive).

So my first question is: is this what we want? Do we want that adding to the collection will possibly not be visible to existing readers?

If the answer is yes, then you might be able to solve this problem by having your class maintain 2 List objects. Whenever you ask for an enumerator, if the 2nd one doesn't match the first (compare size), copy 1st to 2nd, and give an enumerator to the 2nd one. This allows you to add to the 1st, and lazily copy to 2nd when readers want to enumerate. It does mean more allocs, but might still be simpler?

Reply:
Yes, it is important to have one list writing, while others read. It happens a lot when, for example, the log is not completely read yet. The ui already starts drawing the fist page as soon as possible.

Having a second list is possible, but wouldn't this be more complex then the current solution? And the allocs are important, since we are talking about 3 lists per commit. That's also why I initialize the arrays with a size of 2. Usually the lists are not larger then 2. It happens sometimes and then they need to be expanded (=copied).

The code is refactored now, so the default .net enumerator is used. Including all the plumbing (usings, comments, whitespaces), the file is 50 lines.

Reply:
Oh, if it's 50 lines of code, and using the default enumerator now, I'd say it's fine :)

Reply:
Thanks for being critical..:P

Reply:
I meant it positive... it came out a bit wrong, sorry.. ;)